### DISCUSSION TITLE
Writing automated tests for Proxy Contract

### DISCUSSION THREAD

**Discussion by hadoanngoc** [ConsenSys-archive/truffle#5298] [General Q&A]
I'm writing tests for an upgradeable contract. The first run works well. All tests passed.

However, runing again is showing me this error

**Error: Returned error: VM Exception while processing transaction: revert Ownable: caller is not the owner**

It seems like my Contract's owner was transferred to another address during the previous test. So I try to remove the build folder, compile everything again, remove .openzepplin folder, run again. But it still shows the above error.

Any idea about this?

Below is the test (I ignore the "require" lines...
```

describe("Test Upgrade Contracts", function() {
  beforeEach(async function () {
    this.timeout(20000);
  });

  it('deploy V2 and increase current value from V1', async () => {
    const coming = await deployProxy(ComingLock, []);
    let value = await coming.totalLocks();
    console.log("Value before v2: ", value.toString());
    const comingV2 = await upgradeProxy(coming.address, ComingLockV2);
    await comingV2.increaseLocksCount();

    value = await comingV2.totalLocks();
    assert.equal(value.toString(), '2');
  });
});
```

In the first run, I noticed this

```
Compiling your contracts...
===========================
> Everything is up to date, there is nothing to compile.
✔ 0xC89Ce4735882C9F0f0FE26686c53074E09B0D550 (transparent) proxy ownership transfered through admin proxy
```

So somehow the ownership was transfered, but I don't know why it was transfered, and how to reset it

**Comment by cliffoo** [comment]
Hey @hadoanngoc thanks for opening this discussion!

I cannot tell from the snippet you shared, but it looks like somewhere in your
- Deployments
- Tests
- Contract constructors
- Contract interactions subsequent to deployment
The proxy changed owner.

I'm guessing you're using openzepplin here, look for a `changeAdmin(newAdmin)` call somewhere?

Additionally, you may find the `--reset` flag for the `migrate` command useful as you debug.

Feel free to share reproduction steps and I'd be happy to take a deeper look :)

**Comment by hadoanngoc** [comment]
Hi,

Thanks for the reply @cliffoo .

This is "npx truffle test", not "truffle migrate" command. And I do not call "changeAdmin" anywhere. 

I have re-compile all the contracts, restart local server and run "npx truffle test" again. But it seems the proxy ownership keeps changing everytime I run the tests, like this

✘ 0x4cFB3F70BF6a80397C2e634e5bDd85BC0bb189EE (transparent) proxy ownership not affected by admin proxy
✘ 0xc34175A79ACca40392bECD22ff10fAeBFE780Ae7 (transparent) proxy ownership not affected by admin proxy
✔ 0x9561C133DD8580860B6b7E504bC5Aa500f0f06a7 (transparent) proxy ownership transfered through admin proxy

**Reply by cliffoo** [reply]
Hmm could you provide the minimal reproduction steps for this? I can look into it more then. Thanks.

**Reply by hadoanngoc** [reply]
Hi,

I found the issue. I have a migration file which call this command

```
if (network !== 'test') {
    await admin.transferProxyAdminOwnership(gnosisSafe);
}
```

The network to ignore should be "development", not "test". Therefore, during the first run, ownership of the ProxyAdmin contract was transferred. In the next run, default accounts[0] is not the owner of the proxy anymore, so it cannot deploy or upgrade contracts.

A new thing to me is that, during the "npx truffle test", truffer only runs the migration files.

Another point, once ProxyAdmin contract is deployed, it is cached (ie. not deployed again in new tests) and alive in all tests. Therefore, we need to delete this file in /.openzeppelin/unknown-xxxx.json metadata file to force the new deployment of ProxyAdmin. (but I'm sure I have done that several times, don't know why in the last time it works)

Thanks @cliffoo

**Reply by cliffoo** [reply]
Do you have a `Migrations.sol` file that looks like [this](https://github.com/trufflesuite/truffle-init-default/blob/master/contracts/Migrations.sol)? If so that's the reason why your contracts don't re-deploy on subsequent `migrate` calls. And yes you can remove the contract artifact files (the metadata files you're referring to) or do a `truffle migrate --reset` (see more in `truffle help migrate`) to run all migrations.

> A new thing to me is that, during the "npx truffle test", truffer only runs the migration files.

So `truffle test` is not running tests for you? Only migrations? Try making a test file under /test, and run `truffle test` to see if the test runs.

A moot but simple test:
```js
contract("foo", () => {
  it("bar", () => {
    assert.deepEqual(1, 1);
  });
});
```

Read more about testing [here](https://trufflesuite.com/docs/truffle/testing/testing-your-contracts/).

Anyway glad you resolved your issue! Cheers.
